helix-view/src/file_event.rs

```diff
@@ -0,0 +1,206 @@
+
+    pub fn unregister(&self, client_id: usize, registration_id: String) {
+        let _ = self.tx.send(Event::Unregister {
+            client_id,
+            registration_id,
+        });
+    }
+
+    pub fn file_changed(&self, path: PathBuf) {
+        let _ = self.tx.send(Event::FileChanged { path });
+    }
+
+    pub fn remove_client(&self, client_id: usize) {
+        let _ = self.tx.send(Event::RemoveClient { client_id });
+    }
+
+    async fn run(mut rx: mpsc::UnboundedReceiver<Event>) {
+        let mut state: HashMap<usize, ClientState> = HashMap::new();
+        while let Some(event) = rx.recv().await {
+            match event {
+                Event::FileChanged { path } => {
+                    log::debug!("Received file event for {:?}", &path);
+
+                    let mut dropped_clients = Vec::new();
+
+                    let futures = state
+                        .iter()
+                        .filter(|(_, client_state)| {
+                            client_state
+                                .registered
+                                .values()
+                                .any(|glob| glob.is_match(&path))
+                        })
+                        .filter_map(|(client_id, client_state)| {
+                            client_state.client.upgrade().or_else(|| {
+                                dropped_clients.push(*client_id);
+                                None
+                            })
+                        })
+                        .filter_map(|client| {
+                            let uri = lsp::Url::from_file_path(&path).ok()?;
+                            log::debug!(
+                                "Sending didChangeWatchedFiles notification to client '{}'",
+                                client.name()
+                            );
+                            Some(client.did_change_watched_files(vec![lsp::FileEvent {
+                                uri,
+                                // We currently always send the CHANGED state
+                                // since we don't actually have more context at
+                                // the moment.
+                                typ: lsp::FileChangeType::CHANGED,
+                            }]))
+                        });
+                    for res in join_all(futures).await {
+                        if let Err(err) = res {
+                            log::warn!("Failed to send didChangeWatchedFiles notification to client: {err}");
+                        }
+                    }
+
+                    // Remove any clients that've been dropped but we still
+                    // have a weak reference to.
+                    if !dropped_clients.is_empty() {
+                        log::warn!("LSP clients were dropped: {:?}", dropped_clients);
+                        for id in dropped_clients {
+                            state.remove(&id);
+                        }
+                    }
```

Comment:
you can achieve that using `HashMap::retain` (and returning false in that case and true otherwise)

Reply:
If I attempt to use `HashMap::retain` inside of the for loop I get the same `cannot borrow state as mutable because it is also borrowed as immutable` error, since it also requires a mutable reference

Reply:
you can use `HashMap::retain` *instead* of the for loop

Reply:
Ahh okay, sorry I misunderstood! The problem I'm running into now is that the closure that `HashMap::retain` takes cannot be `async` (required to await the LSP notification) 😞

Reply:
👍 that makes sense, just pushed that change!